File: 排序算法/qsort.go
package main

import "fmt"

func partition(a []int, left, right int) int {
	low, up := left+1, right
	x := a[left]
	for {
		for a[low] < x && low < up {
			low++
		}
		for a[up] > x {
			up--
		}
		if low >= up {
			break
		}
		a[low], a[up] = a[up], a[low]
	}
	a[left], a[up] = a[up], a[left]
	return up
}

func quickSort(a []int, left, right int) {
	if left < right {
		mid := partition(a, left, right)
		quickSort(a, left, mid-1)
		quickSort(a, mid+1, right)
	}
}

func main() {
	a := []int{29, 22, 33, 6, 77, 99, 88, 250, 100}
	fmt.Println("before: ", a)
	quickSort(a, 0, len(a)-1)
	fmt.Println("after: ", a)
}

File: 排序算法/qsort.cpp
// 快排C++实现

template <class Type>
void QuickSort(Type a[], int p, int r)
{
    if (p < r)
    {
        int q = Partition(a, p, r);
        QuickSort(a, p, q - 1);
        QuickSort(a, q + 1, r);
    }
}

template <class Type>
int Partition(Type a[], int p, int r)
{
    int i = p, j = r + 1;
    Type x = a[p];
    while (true)
    {
        while (a[++i] < x && i < r);
        while (a[--j] > x);
        if (i >= j)
            break;
        int tmp = a[i];
        a[i] = a[j];
        a[j] = tmp;
    }
    a[p] = a[j];
    a[j] = x;
    return j;
}